tests/build_cache_images.sh

```diff
@@ -0,0 +1,67 @@
+#!/bin/bash
+
+set -e
+
+CACHE_REGISTRY=${1}
+APP_NAME=${2}
+TARGET_NAME=${3}
+DOCKERFILE=${4:-"Dockerfile"}
+
+if [ -z "$CACHE_REGISTRY" ] || [ -z "$APP_NAME" ] || [ -z "$TARGET_NAME" ]; then
+    echo "Usage: build-cache-images.sh cache-registry appname target-name [dockerfile-name]"
+    exit 1
+fi
+
+# cd to the directory of this script
+cd "$(dirname "$(readlink -f "${BASH_SOURCE[0]}")")"
+
+# Ensure the app exists
+APP_DIR="apps/${APP_NAME}"
+if [ ! -d "$APP_DIR" ]; then
+    echo "App $APP_NAME does not exist"
+    exit 1
+fi
+if [ ! -e "$APP_DIR/${DOCKERFILE}" ]; then
+    echo "No Dockerfile (${DOCKERFILE}) found in $APP_NAME folder"
+    exit 1
+fi
+
+function hashdir(){
+    starting_dir="$(pwd)"
+    target_dir="$1"
+    cd "$target_dir"
+
+    array=($(find . -not -type d))
+    readarray -t sorted < <(sort < <(printf '%s\n' "${array[@]}"))
+    result=""
+    for i in ${sorted[@]};do
+        result+="$(sha256sum $i)\n"
+    done
+    cd "$starting_dir"
+    echo -e "$result" | sha256sum | awk '{ print $1 }'
+}
+
+# Compute the hash of the app's files
+HASH=$(hashdir "$APP_DIR")
+echo "HASH: ${HASH:0:10} (${HASH})"
+HASH="${HASH:0:10}"
+CACHE_TAG="${CACHE_REGISTRY}/${APP_DIR}:${DOCKERFILE}-${HASH}"
+
+# Check if the image already exists
+set +e
+docker pull "${CACHE_TAG}"
+EXITCODE=$?
+set -e
```

Comment:
Only check the failure of pull image from cache?

Reply:
We cannot determine the cause of the failure of this command. So we assume that whatever the reason is, the image doesn’t exist in the cache.
If the issue is related to authentication, the push command later would fail too.

Reply:
`set -e` helps here since we want all commands to succeed.

Reply:
No, this specific command may fail   If the image is not in the cache, it should fail and exit with non-0 status code. We capture the status code to check it the image exists.

If it doesn't exist, we build the image.

Reply:
`func pull_from_cache || func build_and_push` then we can use `set -e`.

Reply:
That sounds reasonable. I made the change as you requested